### ISSUE TITLE
why update grad value in DoBoost function (gblinear-inl.hpp)

### ISSUE THREAD

**Issue by EasonZhaoZ** [dmlc/xgboost#433]
``` 
 83       #pragma omp parallel for schedule(static)
 84       for (bst_omp_uint i = 0; i < nfeat; ++i) {
 85         const bst_uint fid = batch.col_index[i];
 86         ColBatch::Inst col = batch[i];
 87         for (int gid = 0; gid < ngroup; ++gid) {
 88           double sum_grad = 0.0, sum_hess = 0.0;
 89           for (bst_uint j = 0; j < col.length; ++j) {
 90             const float v = col[j].fvalue;
 91             bst_gpair &p = gpair[col[j].index * ngroup + gid];
 92             if (p.hess < 0.0f) continue;
 93             sum_grad += p.grad * v;
 94             sum_hess += p.hess * v * v;
 95           }
 96           float &w = model[fid][gid];
 97           bst_float dw = static_cast<bst_float>(param.learning_rate *
 98                                                 param.CalcDelta(sum_grad, sum_hess, w));
 99           w += dw;
100           // update grad value  这里有必要吗？
101           for (bst_uint j = 0; j < col.length; ++j) {
102             bst_gpair &p = gpair[col[j].index * ngroup + gid];
103             if (p.hess < 0.0f) continue;
104             p.grad += p.hess * col[j].fvalue * dw;
105           }
106         }
107       }
108     }
```

**Comment by tqchen** [comment]
Because we are doing multiple iterations of parameter updating in this loop, updating the gradient value is needed
